### PR TITLE
Solvent as sample

### PR DESCRIPTION
**Pull request by JasonYCHuang**
Several things are done at this PR.
~~1. fix rspec reaction_api_spec~~
2. fix some react warning
3. Solvent as sample
4. Report add reactions status
5. Refactor reaction scheme table

### CODE REVIEW

**Comment by Sergeykot** [comment]
conflicts probably in schema.rb

**Comment by PiTrem** [comment]
conflict in reaction_api_spec because...

**Comment by JasonYCHuang** [comment]
I will drop my reaction_api_spec and fix it in another branch.

File: app/assets/javascripts/components/Material.js [hunk 1]
```diff
@@ -106,86 +106,19 @@ export default class Material extends Component {
...
-        <Button
-          bsStyle="danger"
-          onClick={() => deleteMaterial(material)} >
-          <i className="fa fa-trash-o"></i>
-        </Button>
-      </td>
-    </tr>
+    return (
+      this.props.materialGroup !== 'solvents'
+      ?
```
Comment (Sergeykot) [inline app/assets/javascripts/components/Material.js]: maybe if-else instead?

File: app/assets/javascripts/components/Material.js [hunk 2]
```diff
@@ -201,13 +134,36 @@ export default class Material extends Component {
         <NumeralInputWithUnitsCompo
           precision={4}
           value={material.equivalent}
-          disabled={material.reference && material.equivalent}
+          disabled={(material.reference && material.equivalent) !== false}
```
Comment (Sergeykot) [inline app/assets/javascripts/components/Material.js]: why !== false

File: app/assets/javascripts/components/MaterialGroup.js [hunk 3]
```diff
@@ -38,31 +42,70 @@ export default class MaterialGroup extends Component {
...
-          <tbody>
-            {contents.map(function(item) {
-              return item;
-            })}
-          </tbody>
-        </table>
-      </div>
-    );
+      materialGroup === 'solvents'
+      ?
```
Comment (Sergeykot) [inline app/assets/javascripts/components/MaterialGroup.js]: if-else maybe ?

File: app/assets/javascripts/components/report/ReactElements.js [hunk 4]
```diff
@@ -28,7 +75,10 @@ const MaterialContent = ({show, starting_materials, reactants, products}) => {
   }
 
   const rows = (material, isProduct) => {
-    return material.map((sample, i) => {
+    return (
+      typeof material != 'undefined'
+      ?
```
Comment (Sergeykot) [inline app/assets/javascripts/components/report/ReactElements.js]: I don't like this style

Comment (Sergeykot) [inline app/assets/javascripts/components/report/ReactElements.js]: if-else would be more readable

File: app/assets/javascripts/components/report/ReactElements.js [hunk 5]
```diff
@@ -79,6 +131,59 @@ const MaterialContent = ({show, starting_materials, reactants, products}) => {
...
+      solventsData()
+      :
+      solventData()
+    )
+  }
+
+  return (
+    show
+    ?
```
Comment (Sergeykot) [inline app/assets/javascripts/components/report/ReactElements.js]: if-else

File: app/models/sample.rb [hunk 6]
```diff
@@ -280,6 +284,18 @@ def amount_ml type = 'target'
...
   end
 
+  def preferred_label
+    self.external_label.present? ? self.external_label : self.molecule.iupac_name
+  end
+
+  def preferred_tag
+    tag = self.preferred_label
+    return tag && tag.length > 20 ?
+              tag[0, 20] + "..."
```
Comment (Sergeykot) [inline app/models/sample.rb]: if-else

File: lib/report/docx/image.rb [hunk 7]
```diff
@@ -45,8 +46,11 @@ def materials_svg_paths
         return paths
       end
 
-      def solvent_svg_paths
-        [obj.solvent].reject{|c| c.blank?}.join(", ")
+      def solvents
+        obj.solvents.present? ?
+          obj.solvents.map{ |s| s.preferred_tag }
```
Comment (Sergeykot) [inline lib/report/docx/image.rb]: if-else

File: app/assets/javascripts/components/Material.js (hunk 2 above)
Comment (JasonYCHuang) [inline app/assets/javascripts/components/Material.js]: ```material.reference``` coulde be 1
```material.equivalent```  coulde be 1

```1 && 1 -> 1```

We need boolean for disabled, but sometimes ```{material.reference && material.equivalent}```  returns 1 and React.js will warn you.

**Pull request merged**